src/primitive.ts

```diff
@@ -36,141 +36,167 @@ let LEADING_ZERO = /^[+-]?0[0-9_]/
 }
 
-export function parseValue (value: string, toml: string, ptr: number): boolean | number | TomlDate {
-	// Constant values
-	if (value === 'true') return true
-	if (value === 'false') return false
-	if (value === '-inf') return -Infinity
-	if (value === 'inf' || value === '+inf') return Infinity
-	if (value === 'nan' || value === '+nan' || value === '-nan') return NaN
-
-	if (value === '-0') return 0 // Avoid FP representation of -0
-
-	// Numbers
-	let isInt
-	if ((isInt = INT_REGEX.test(value)) || FLOAT_REGEX.test(value)) {
-		if (LEADING_ZERO.test(value)) {
-			throw new TomlError('leading zeroes are not allowed', {
-				toml: toml,
-				ptr: ptr
-			})
-		}
-
-		let numeric = +(value.replace(/_/g, ''))
-		if (isNaN(numeric)) {
-			throw new TomlError('invalid number', {
-				toml: toml,
-				ptr: ptr
-			})
-		}
-
-		if (isInt && !Number.isSafeInteger(numeric)) {
-			throw new TomlError('integer value cannot be represented losslessly', {
-				toml: toml,
-				ptr: ptr
-			})
-		}
-
-		return numeric
-	}
-
-	let date = new TomlDate(value)
-	if (!date.isValid()) {
-		throw new TomlError('invalid value', {
-			toml: toml,
-			ptr: ptr
-		})
-	}
-
-	return date
+export function parseValue(
+  value: string,
+  toml: string,
+  ptr: number,
+): boolean | number | TomlDate | BigInt {
+  // Constant values
+  if (value === 'true') return true
+  if (value === 'false') return false
+  if (value === '-inf') return -Infinity
+  if (value === 'inf' || value === '+inf') return Infinity
+  if (value === 'nan' || value === '+nan' || value === '-nan') return NaN
+
+  if (value === '-0') return 0 // Avoid FP representation of -0
+
+  // Numbers
+  let isInt
+  if ((isInt = INT_REGEX.test(value)) || FLOAT_REGEX.test(value)) {
+    if (LEADING_ZERO.test(value)) {
+      throw new TomlError('leading zeroes are not allowed', {
+        toml: toml,
+        ptr: ptr,
+      })
+    }
+
+    // Remove underscores for processing
+    const cleanValue = value.replace(/_/g, '')
+
+    // Check if it's an integer with more than 15 digits (exceeds safe integer precision)
+    if (isInt && /^[+-]?\d{16,}$/.test(cleanValue)) {
+      try {
+        // Use BigInt for large integers
+        return BigInt(cleanValue)
+      } catch {
+        throw new TomlError('invalid number', {
+          toml: toml,
+          ptr: ptr,
+        })
+      }
+    } else {
+      // Use regular number for smaller integers and floats
+      let numeric = +cleanValue
+      if (isNaN(numeric)) {
+        throw new TomlError('invalid number', {
+          toml: toml,
+          ptr: ptr,
+        })
+      }
+
+      if (isInt && !Number.isSafeInteger(numeric)) {
+        try {
+          // Fall back to BigInt if the number is too large for safe integers
+          return BigInt(cleanValue)
+        } catch {
+          throw new TomlError('integer value cannot be represented losslessly', {
+            toml: toml,
+            ptr: ptr,
+          })
+        }
+      }
```

Comment:
I'm very, very unsure about whether this is a desirable default behavior. This effectively turns all `number` into `number | bigint` which can be quite a pain to deal with (with strict type safety compliance that is). Plus that's a breaking change as it changes the types that can be received.

It'd be best to have an option to opt into this behavior, or perhaps (what I was initially thinking) an option that makes all integers parsed as bigints, for the sake of consistency

Reply:
I think a good solution would be 3 modes :
- ints as number and error if not possible without loss of info
- ints as number and bigint if not possible without loss of info
- ints as bigints, always

I'm relatively confident I could help with typing.